desktop_version/src/Graphics.cpp

```diff
@@ -3000,15 +2999,22 @@ void Graphics::reloadresources(bool fast /*= false*/) {
     grphx.init();
     pre_fakepercent.store(90);
 
-    images.clear();
-    tiles.clear();
-    tiles2.clear();
-    tiles3.clear();
-    entcolours.clear();
-    sprites.clear();
-    flipsprites.clear();
-    tele.clear();
-    customtiles.clear();
+#define SDLVecClear(A) for(auto x : A){ SDL_FreeSurface(x); } A.clear();
+#define SDLMapClear(A) for(auto x : A){ for(auto y : x.second){ SDL_FreeSurface(y); } } A.clear();
+    SDLVecClear(images);
+    SDLVecClear(tiles);
+    SDLVecClear(tiles2);
+    SDLVecClear(tiles3);
+    SDLVecClear(entcolours);
+    SDLVecClear(sprites);
+	// Freeing flipsprites causes double free
+	// Someone who understands this please deal with this
+	flipsprites.clear();
+    SDLVecClear(tele);
+    SDLMapClear(customtiles);
+	SDLMapClear(customsprites);
+#undef SDLVecClear
+#undef SDLMapClear
```

Comment:
This should be upstreamed (except for the `auto` part. And if you have to use `auto` in order to use the macros at all, might as well just remove the macros and expand them out).

I'll look at the `flipsprites` double-free later.

Reply:
Scratch that upstreaming, vanilla doesn't have `Graphics::reloadresources()`.

Reply:
Ok, so as far as I can tell, it's not a double-free - it's simply an invalid pointer that somehow got pushed to the vector.

Reply:
Ah, no, it *is* a double-free after all.

In `Graphics::MakeSpriteArray()`:
```c++
    if ((sprites_height / 32) < default_rows) {
        int rest = default_rows - (sprites_height / 32);
        for (int j = 0; j < rest; j++) {
            for(int i = 0; i < 12; i++) {
                SDL_Surface* temp = GetSubSurface(grphx.im_sprites,0*32,4*32,32,32);
                sprites.push_back(temp);
                flipsprites.push_back(temp);
            }
        }
    }
```
Looks like instead of handling an expanded `flipsprites.png` separately, we just push the expanded `sprites.png` sprites onto `flipsprites` instead of having separate `flipsprites` sprites. Thus, some surfaces have already been freed by the time we get around to freeing the entire `flipsprites` vector, causing a double-free.

Reply:
(Btw, @AllyTally is responsible for this state of affairs, `git blame` points to 20dc40d0d769c38041cbed689e479411c4865a91.)